### DISCUSSION TITLE
Detailed error logging

### DISCUSSION THREAD

**Discussion by zakkor** [pocketbase/pocketbase#2420] [Q&A] [answered]
I'd like to set up global error logging, including full stack traces created by using `errors.New` from `"github.com/pkg/errors"`.

```
defaultErrHandler := e.Router.HTTPErrorHandler
e.Router.HTTPErrorHandler = func(c echo.Context, err error) {
	if c.Response().Committed {
		return
	}

	if err != nil {
		fmt.Printf("%+v\n", err)
	}

	defaultErrHandler(c, err)
}
```

```
func postsBeforeCreate(e *core.ModelEvent) error {
	return errors.New("test error")
}
```

```
app.OnModelBeforeCreate("posts").Add(postsBeforeCreate)
```

However, in this handler, I only receive generic errors like `Failed to create record.`, and the actual error is swallowed away somewhere. It also does not seem to be wrapped, as errors.Unwrap(err) returns nil.

Any tips?

**Comment by ganigeorgiev** [comment, marked as answer]
Request errors are usually returned as [`*apis.ApiError`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.15.2/apis#ApiError) (it implements the `error` interface), so you can try to type assert it (eg. `err.(*apis.ApiError)`) to access its fields.

The original err (or any other error meta data related to the ApiError) could be accessed via `ApiError.RawData()` (note that it could be anything - `error` value, map, text, etc.).

**Reply by zakkor** [reply]
Works perfectly!

@ganigeorgiev Follow-up question:
What happens to errors returned from `app.OnModelAfter*` handlers?
They're not part of the HTTP response, so they won't be given to the HTTP error handler middleware.
But I'd like to handle those as well - as far as I can tell, they are silently swallowed, which seems like a default behaviour that's not super ideal.

**Reply by ganigeorgiev** [reply]
> What happens to errors returned from app.OnModelAfter* handlers?

Currently they are only logged when in debug mode. This may change with the planned hooks refactoring and logs generalization.

**Reply by pnmcosta** [reply]
@ganigeorgiev could we have a bit more of an insight into the planned "hooks refactoring and logs generalization."?

**Reply by ganigeorgiev** [reply]
I'm still experimenting with the hooks refactoring and don't have yet a finalized proposal (_my idea was initially to merge the before/after hooks into one with a chainable `resolver()` argument to allow sharing data between the before/after events but this didn't work well with the `OnModel*` hooks and the Dao transactions_).

About the logs generalization - the main goal is to allow custom app level logs that will be stored in the `pb_data/logs.db`. Currently we store only request logs, so we'll have to also change the db logs structure and the Admin UI visualization (this for now is planned probably for v0.18/v0.19).

**Reply by pnmcosta** [reply]
Got it, thank you.

Regarding hooks, I think the Before/After works quite well, although you do need to dig into the PB code to figure out a couple of things - I personally don't mind it - from my limited experience:

1. Not all After hooks affect the response, for example returning an error on `OnRecordAfterAuthWithOAuth2Request` is just consumed to logs on debug.
2. With `OnRecordBeforeAuthWithOAuth2Request` I wish I could append/modify the CreateData, unfortunately cause it gets copied to the form early, modifying the request data then has no effect. Could `CreateData` be part of the hook event struct?

Regarding application logging, it's something high on my pre production list, if we can surface app logs on the admin UI too that would be awesome - I wouldn't need to find a logging provider/alternative. Please do consider expanding the new Backup feature to the logs db too, with separate schedules.

Thank you once again for all the hard work 💪

**Reply by ganigeorgiev** [reply]
Thank you for the feedback! I'll consider it during the refactoring.
